crd-generator/src/main/java/io/strimzi/crdgenerator/Property.java

```diff
@@ -274,25 +283,54 @@ static Map<Class<?>, String> subtypeMap(Class<?> crdClass) {
         }
     }
 
-    static List<Class<?>> subtypes(Class<?> crdClass) {
+    /**
+     * This method is used to provide the typed classes that are part of a typed API. This method is used to generate
+     * the CRDs and the API reference docs.
+     *
+     * @param crApiVersion  Strimzi API version for which the type classes should be returned (some types might be present only in some versions)
+     * @param crdClass      The parent class for which the type identifiers should be returned
+     *
+     * @return  List of the available subtype classes
+     */
+    static List<Class<?>> subtypes(ApiVersion crApiVersion, Class<?> crdClass) {
         JsonSubTypes subtypes = crdClass.getAnnotation(JsonSubTypes.class);
         if (subtypes != null) {
             ArrayList<Class<?>> result = new ArrayList<>(subtypes.value().length);
             for (JsonSubTypes.Type type : subtypes.value()) {
-                result.add(type.value());
+                // When the Strimzi API version is null, we return all types. This is usually used for validations and docs.
+                // When the PresentInVersion annotation is missing, the type is included in all versions
+                // When it is present, it is included only in matching versions
+                if (crApiVersion == null
+                        || !type.value().isAnnotationPresent(PresentInVersions.class)
+                        || ApiVersion.parseRange(type.value().getAnnotation(PresentInVersions.class).value()).contains(crApiVersion)) {
+                    result.add(type.value());
+                }
             }
             return result;
         } else {
             return emptyList();
         }
     }
 
-    static List<String> subtypeNames(Class<?> crdClass) {
+    /**
+     * This method is used to provide the type identifiers for typed APIs. This method is used to generate the CRDs.
+     *
+     * @param crApiVersion  Strimzi API version for which the identifiers should be returned (some types might be present only in some versions)
+     * @param crdClass      The parent class for which the type identifiers should be returned
+     *
+     * @return  List of the available types
+     */
+    static List<String> subtypeNames(ApiVersion crApiVersion, Class<?> crdClass) {
         JsonSubTypes subtypes = crdClass.getAnnotation(JsonSubTypes.class);
         if (subtypes != null) {
             ArrayList<String> result = new ArrayList<>(subtypes.value().length);
             for (JsonSubTypes.Type type : subtypes.value()) {
-                result.add(type.name());
+                // When the PresentInVersion annotation is missing, the type is included in all versions
+                // When it is present, it is included only in matching versions
+                if (!type.value().isAnnotationPresent(PresentInVersions.class)
```

Comment:
can't `crApiVersion` be null in this case as well? as it's checked for `subtypes` method?

Reply:
@ppatierno  No, the `crApiVersion ` should not be null here. It migth be used with the null value only in places where we want to get the things regardless the version. That is in Docs in some cases and in the CRD generator only when we just want to get the typed classes, regardless of the version (as that will be checked later). But not in this method.

If you want, I can add an explicit null check. But it is basically used only internally, so not sure having an NPE here is a problem. Maybe clarifying it in the Javadoc would ne enough?

Reply:
Yeah ok thanks, let's just clarify in the Javadoc.